Iptables functions improvement

arnova opened the pull request.

abelbeck commented on share/arno-iptables-firewall/environment:
```diff
@@ -320,18 +320,25 @@ ip4tables()
 
   if [ $retval -ne 0 ]; then
     # Show any (error) messages in red
-    printf "\033[40m\033[1;31m${IP4TABLES} $*\nERROR ($retval): ${result}\033[0m\n" >&2
```
Wouldn't simply renaming `result` -> `err_result` be the same as all the new code ?

arnova replied:
No ;-) The idea is that we no longer redirect stderr to stdout and then process only stdout (which is too error prone, as it turns out). So now we just pass through stdout (via &3) and then redirect stderr to stdout and process then only stderr. So it's certainly different.

abelbeck replied:
No `:-)` (I wasn't clear) I was only referring to the one red line before this comment and the 10 green lines after this comment ... except for a added newline, the output looks the same to me.

arnova replied:
Well the idea is that when iptables returns 0 ("success"), its stderr still gets send to stderr instead of stdout (like we previously did). Have a close look at the usages of retval here.

abelbeck replied:
Ahh, I see it now (sorry for the delay `:-)`)

Possibly this is less change and more clear ... you decide:
```
  if [ $retval -ne 0 ]; then
    # Show any (error) messages in red
    printf "\033[40m\033[1;31m${IP4TABLES} $*\nERROR ($retval): ${err_result}\033[0m\n" >&2
    if note_iptables_error "$@"; then
      RULE_WARNING=$((RULE_WARNING + 1))
    fi
  elif [ -n "$err_result" ]; then
    # Success, but with stdout
    echo "$err_result" >&2
  fi
```

arnova replied:
Yes, I think you're right. That looks cleaner :-)

abelbeck commented on share/arno-iptables-firewall/environment:
```diff
@@ -320,18 +320,25 @@ ip4tables()
+  exec 3>&1
+  err_result=`$IP4TABLES${IPTABLES_OPTIONS:+ $IPTABLES_OPTIONS} "$@" 2>&1 1>&3 |grep -v -e 'WARNING:.*match is obsolete' -e 'iptables-legacy tables present'`
   retval=$?
 
```
I wonder if we want to add:
```
exec 3>&-
```
to close fd3 ?

Ref: https://tldp.org/LDP/abs/html/io-redirection.html

arnova replied:
Yes, we should. Forgot to cp that from aif-job-processor

abelbeck commented on share/arno-iptables-firewall/environment:
```diff
@@ -320,18 +320,25 @@ ip4tables()
 {
   local result retval IFS=' '
```
```
local err_result retval IFS=' '
```

arnova commented:
@abelbeck : Updated with your suggestions.

abelbeck commented on share/arno-iptables-firewall/environment:
```diff
@@ -381,15 +405,23 @@ ip4tables_restore()
+  exec 3>&1
+  err_result=`$IP4TABLES_RESTORE "$@" 2>&1 1>&3`
   retval=$?
 
```
Similar fixes apply to `ip4tables_restore()` and `ip6tables_restore()`

arnova replied:
Well the logic there was always a little different since stdout there gets redirected by the caller. I'm not sure whether we should change this. Currently when an error occurs all output (stdout+stderr) gets sent to stderr so the user knows what's going on. What we could do is in case of an error also send stdout to stderr, so the caller still gets stdout to redirect. I hope your understand what I mean.

arnova commented:
@abelbeck : Updated once again. Also updated ip4tables_save & ip6tables_save to make them in line with the rest.

abelbeck commented:
> @abelbeck : Updated once again. Also updated ip4tables_save & ip6tables_save to make them in line with the rest.

I think ip4tables_save & ip6tables_save should be left alone, plus you missed a couple things.

I'm worried redirecting to fd3 and then closing ... simpler to leave unchanged.

arnova commented:
Ok, I've reverted the changes for ip4|6tables_save. Do note that closing fd3 shouldn't cause any issues with redirection and such. I've also adjusted the changes in ip4tables/ip6tables since using grep directly (obviously) screws up retval. If you're fine with the changes as-is, hit green please :-)

abelbeck commented:
@arnova : I think 0dd5afae5e981ed03739057dfb99c348985cf33d needs to be reverted, as it messes with the return value.

arnova commented:
> @arnova : I think [0dd5afa](https://github.com/arno-iptables-firewall/aif/commit/0dd5afae5e981ed03739057dfb99c348985cf33d) needs to be reverted, as it messes with the return value.

Crap, you're right. Nice catch. Any ideas for a proper fix?

abelbeck commented:
> Any ideas for a proper fix?

Revert it `:-)`

Is adding an INDENT important ?

arnova commented:
> > Any ideas for a proper fix?
> 
> Revert it `:-)`
> 
> Is adding an INDENT important ?

Don't know, I never use those functions. I thought you wrote them back in the days (or was it Phillip?)...

arnova commented:
> > > Any ideas for a proper fix?
> > 
> > 
> > Revert it `:-)`
> > Is adding an INDENT important ?
> 
> Don't know, I never use those functions. I thought you wrote them back in the days (or was it Phillip?)...

Just had a look: it seems like those functions never output anything to stdout (unless -v is used, which is not the case here) so using INDENT seems useless.

abelbeck commented:
> I thought you wrote them back in the days

Yes its mine, this allowed much faster blocklists ... but with IPTABLES_IPSET=1 this code is no longer used.

arnova commented:
@abelbeck : Updated once again :/
